File: flutter/test/models/vibration_pattern_test.dart
import 'package:careconnect_mobile/models/vibration_pattern.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('VibrationPattern', () {
    test('covers the three alert types the prototype names', () {
      expect(
        VibrationPattern.values.map((VibrationPattern p) => p.alertType),
        <String>['Appointment', 'Medication', 'Missed / Escalated'],
      );
      expect(VibrationPattern.appointment.rhythmName, 'Long-short-long');
      expect(VibrationPattern.medication.rhythmName, 'Double pulse');
      expect(VibrationPattern.missed.rhythmName, 'Rapid burst');
    });

    test('every pattern is printed as well as felt', () {
      for (final VibrationPattern pattern in VibrationPattern.values) {
        expect(pattern.glyph, isNotEmpty);
        expect(pattern.semanticLabel, contains(pattern.alertType));
        expect(pattern.semanticLabel, contains(pattern.rhythmName));
      }
    });

    test('every pattern starts with a vibration and has an odd pulse count',
        () {
      // Pulses alternate vibrate/pause starting with a vibrate, so a
      // well-formed pattern always ends on a vibrate too.
      for (final VibrationPattern pattern in VibrationPattern.values) {
        expect(pattern.pulses, isNotEmpty);
        expect(pattern.pulses.length.isOdd, isTrue,
            reason: '${pattern.name} should not end on a pause');
        expect(pattern.pulses.every((int ms) => ms > 0), isTrue);
      }
    });

    test('the three rhythms are actually distinguishable from each other', () {
      final Set<int> lengths = VibrationPattern.values
          .map((VibrationPattern p) => p.vibrateDurations.length)
          .toSet();
      expect(lengths.length, VibrationPattern.values.length);
    });

    test('duration is the sum of every pulse and pause', () {
      expect(
        VibrationPattern.appointment.duration,
        const Duration(milliseconds: 1220),
      );
    });

    test('no pattern runs long enough to be missed as a buzz', () {
      for (final VibrationPattern pattern in VibrationPattern.values) {
        expect(pattern.duration, lessThan(const Duration(seconds: 2)));
      }
    });

    test('vibrateDurations returns only the vibrating halves', () {
      expect(
        VibrationPattern.appointment.vibrateDurations,
        <int>[400, 120, 400],
      );
    });
  });
}

File: mobile/src/models/__tests__/vibrationPattern.test.ts
import {
  patternDuration,
  patternSemanticLabel,
  vibrateDurations,
  vibrationPatterns,
} from '../vibrationPattern';

describe('vibrationPatterns', () => {
  it('covers the three alert types the prototype names', () => {
    expect(vibrationPatterns.map((p) => p.alertType)).toEqual([
      'Appointment',
      'Medication',
      'Missed / Escalated',
    ]);
    expect(vibrationPatterns.map((p) => p.rhythmName)).toEqual([
      'Long-short-long',
      'Double pulse',
      'Rapid burst',
    ]);
  });

  it('prints every pattern as well as playing it', () => {
    for (const pattern of vibrationPatterns) {
      expect(pattern.glyph.length).toBeGreaterThan(0);
      expect(patternSemanticLabel(pattern)).toContain(pattern.alertType);
      expect(patternSemanticLabel(pattern)).toContain(pattern.rhythmName);
    }
  });

  it('every pattern starts and ends on a vibrate', () => {
    // Pulses alternate vibrate/pause starting with a vibrate, so a well-formed
    // pattern always has an odd number of entries.
    for (const pattern of vibrationPatterns) {
      expect(pattern.pulses.length % 2).toBe(1);
      expect(pattern.pulses.every((ms) => ms > 0)).toBe(true);
    }
  });

  it('the three rhythms are actually distinguishable from each other', () => {
    const shapes = new Set(vibrationPatterns.map((p) => vibrateDurations(p).length));
    expect(shapes.size).toBe(vibrationPatterns.length);
  });

  it('duration is the sum of every pulse and pause', () => {
    expect(patternDuration(vibrationPatterns[0])).toBe(1220);
  });

  it('no pattern runs long enough to be missed as a buzz', () => {
    for (const pattern of vibrationPatterns) {
      expect(patternDuration(pattern)).toBeLessThan(2000);
    }
  });

  it('vibrateDurations returns only the vibrating halves', () => {
    expect(vibrateDurations(vibrationPatterns[0])).toEqual([400, 120, 400]);
  });
});
